Salmon and Transcriptome Index Tests Documentation (#11)
AlexsLemonade/refinebio-processor-documentation: salmon.md

# Salmon

## Overall Tests

### `test_salmon`

Tests the entire salmon pipeline

test steps:
- calls `prepare_job`
    - creates all prerequisites for the salmon pipeline to be run including:
        - ProcessorJob
        - Sample
        - OriginalFiles
        - OrganismIndex
- passes the created base job to `salmon.salmon()` which runs the entire pipeline
- asserts that the `job.success` is true
- asserts that the `sample.is_processed` is false
- asserts that the index type for the organism is 'TRANSCRIPTOME_SHORT'

### `test_no_salmon_on_geo`

Confirms that salmon won't be run on GEO data

test steps:

> - calls `prepere_job`

Reviewer: ```suggestion
- calls `prepare_job`
```

    - creates all prerequisites for the salmon pipeline to be run including:
        - ProcessorJob
        - Sample
        - OriginalFiles
        - OrganismIndex
- creates copies for all original files since they will be cleaned up when the job fails
- edits the sample created in `prepare_job` to have a `source_database` of GEO
- calls `salmon.salmon()` on the base job to run the entire pipeline
- asserts that job context is created
- asserts `job.success` is false
- asserts the `job.failure_reason` is correct
- asserts that the job is not set to retry
- asserts that the original files got cleaned up properly

### `test_salmon_dotsra`

Tests the entire pipeline with a sra file

test steps:
- calls `prepare_dotsra_job`
    - creates all prerequisites for the salmon pipeline to be run including:
        - ProcessorJob
        - Sample
        - OriginalFiles
        - OrganismIndex

>     - uses an RSA file when creating the OriginalFile

Reviewer: ```suggestion
    - uses an SRA file when creating the OriginalFile
```

I think?

- passes the created base job to `salmon.salmon()` which runs the entire pipeline
- asserts that the `job.success` is true

### `test_salmon_dotsra_bad`

Confirms that the job fails with an SRA file that doesn't exist

test steps:
- calls `prepare_dotsra_job` with a file that doesn't exist
    - creates all prerequisites for the salmon pipeline to be run including:
        - ProcessorJob
        - Sample
        - OriginalFiles
        - OrganismIndex

>     - uses an RSA file when creating the OriginalFile

Reviewer: Same comment as above
```suggestion
    - uses an SRA file when creating the OriginalFile
```

- passes the created base job to `salmon.salmon()` which runs the entire pipeline
- asserts that `job_context` was created
- asserts that the `job.success` is false

### `test_salmon_quant_one_sample_double_reads`

> Tests salmon quant on a single sample with two original files (double reads)

Reviewer: I think this is what this means / how I would talk about it
```suggestion
Tests salmon quant on a single sample with two original files (paired end)
```

test steps:
- calls `prepare_organism_indices`
    - creates a dummy OrganismIndex that can be used in the salmon pipeline
- creates the test sample and experiment
- adds an additional fake sample to the experiment to prevent tximport from running on the experiment
    - tximport will only run after salmon quant is run on every sample associated with the experiment
- creates two original files and adds them to the test sample
- calls `salmon._prepare_files` to prepare the original files before running salmon quant
- calls `check_salmon_quant` to run salmon quant and tximport
    - asserts quant output file is made
    - asserts quant output file strongly correlates to reference file

>     - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.

Reviewer: Adding what those columns in the quant file are
```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

- asserts that the test experiment is not ready for tximport yet

### `test_salmon_quant_two_samples_single_read`

Tests salmon quant on two samples each with one original file (single read) that belong to the same experiment

test steps:
- calls `prepare_organism_indices`
    - creates a dummy OrganismIndex that can be used in the salmon pipeline
- creates the test experiment
- creates sample1 and og_file_1, associates the two, and adds the sample to the experiemnt
- creates sample2 and og_file_2, associates the two, and adds the sample to the experiemnt
- calls `salmon._prepare_files()`, passing in  og_file_1
- calls `check_salmon_quant` to run salmon quant and tximport
    - asserts quant output file is made
    - asserts quant output file strongly correlates to reference file

>     - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.

Reviewer: ```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

> - asserts that the experiment is not ready for tximport yet - salmon qunat still needs to be run on sample2

Reviewer: ```suggestion
- asserts that the experiment is not ready for tximport yet - salmon quant still needs to be run on sample2
```

- asserts that there are no tximport outputs yet

> - calls `salmon._prepare_files()`, passing in  og_file_2

Reviewer: ```suggestion
- calls `salmon._prepare_files()`, passing in og_file_2
```

- cleans up tximport data from previous runs if it exists
- calls `check_salmon_quant` to run salmon quant and tximport
    - asserts quant output file is made
    - asserts quant output file strongly correlates to reference file

>     - the files are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a spearman correlation >= 0.99.

Reviewer: ```suggestion
    - the files are strongly correlated if both columns #3 and #4 (zero-indexed; TPM and NumReads) of the two files have a spearman correlation >= 0.99.
```

    - at this point tximport should be run
- asserts that the rds file (tximport output file) has been made
- runs test_tximport.R to verify the output file
- assert that two output files were created

> - assert that the two outut files exist

Reviewer: ```suggestion
- assert that the two output files exist
```

### `test_get_tximport_inputs`

> Tests that tximport only considers RNA-Seq samples from GEO

Reviewer: I would expect this to be SRA not GEO? https://github.com/AlexsLemonade/refinebio-processor-documentation/pull/11/files#diff-76b3c308d473d4219815890b7c1da498a41bd8124fce60ff30938fc6dfe03e6eR21

Author: Yeah I think that's correct - this is incorrectly documented in the refinebio test file as well

- creates dummy `job_context`
- creates dummy Sample
- sets the yml file to one that doesn't exist
- tries to run SalmonTools
- asserts that the job's success is false
- asserts that the job's failure reason is set correctly
- resets the yml file to the correct file

## EarlyTximport Tests

### `test_early_tximport`

Tests that running tximport early works correctly

test steps:
- creates lists of accession codes for incomplete and complete samples
    - incomplete samples has 5 accession codes
    - complete samples has 20 accessioncodes
    - By having 20/25 samples complete, we're just past both the numerical and percent cutoffs for running tximport early
- calls `run_tximport_at_progress_point`
    - creates a dummy `job_context` that could be passed into the salmon pipeline and have all steps related to tximport run successfully.
    - creates a test experiment
    - creates test samples and their associations for incomplete and complete samples and adds them to the test experiment
    - runs tximport
- asserts that 'tximported' is in the final `job_context`
- for each sample in the completed samples list:
    - check that the database associations were made properly 

>     - check that the tpm and rds files were created for the sample

Reviewer: ```suggestion
    - check that the tsv and rds files were created for the sample
```
I believe

# Salmon

## Utils

These functions are used by the tests. Most utils are common code for setting up tests or asserting after the test is run.

### `prepare_organism_indices`

This function creates a base organism index and all of its relations which will be used by salmon in the job pipeline.

### `prepare_job`

This function creates all the prerequisites for the transcriptome index pipeline to be run.

These prerequisites include:

- ProcessorJob
- Sample
- OriginalFiles

### `prepare_dotsra_job`

This function is the same as prepare_job, but the Original File is sra.

### `identical_checksum`

Checks if two files have identical checksums.

### `strong_quant_correlation`

Checks if the two inputed quant files are strongly correlated.

> They are strongly correlated if both columns #3 and #4 (zero-indexed) of the two files have a correlation >= 0.99.

Reviewer 2: This is a very useful level of detail, however it's a bit too far from the tests that actually use it.

- calls `check_salmon_quant`
    - asserts quant output file is made
    - asserts quant output file strongly correlates to reference file
- asserts that the test experiment is not ready for tximport yet

### `test_salmon_quant_two_samples_single_read`

Tests salmon quant on two samples each with one original file (single read) that belong to the same experiment

test steps:
- calls `prepare_organism_indices`
- creates the test experiment
- creates sample1 and og_file_1, associates the two, and adds the sample to the experiemnt
- creates sample2 and og_file_2, associates the two, and adds the sample to the experiemnt
- calls `salmon._prepare_files()`, passing in  og_file_1
- calls `check_salmon_quant`
    - asserts quant output file is made
    - asserts quant output file strongly correlates to reference file
- asserts that the experiment is not ready for tximport yet - salmon qunat still needs to be run on sample2
- asserts that there are no tximport outputs yet
- calls `salmon._prepare_files()`, passing in  og_file_2
- cleans up tximport data from previous runs if it exists
- calls `check_salmon_quant`
    - asserts quant output file is made

>     - asserts quant output file strongly correlates to reference file

Reviewer 2: ```suggestion
    - asserts quant output file strongly correlates to reference file. This is determined to be the case if both columns #3 and #4 (zero-indexed) of the two files have a correlation >= 0.99.
```

Including the detail about what's being tested here means we won't also have to describe it above, and then all the information a scientist might want to know is in one place.  (Although they may also want to know what correlation is being used, e.g., spearman, pearson)

test steps:
- creates dummy `job_context` with both `input_file_path` and `input_file_path_2`
- creates dummy sample and adds it to the `job_context`
- runs SalmonTools
- asserts `job.success` is true
- unpacks the output file
- asserts that both output files match checksums with the expected output files

### `test_single_read`

Tests SalmonTools on a sample with only one OriginalFile (single read)

test steps:
- creates dummy `job_context` with just `input_file_path`
- creates dummy sample and adds it to the `job_context`
- runs SalmonTools
- asserts `job.success` is true
- unpacks the output file
- asserts that the output file's checksum matches with the expected output files'

## DetermineIndexLength Tests

### `test_salmon_determine_index_length_single_read`

Tests that the correct index length is calculated when a sample has one OriginalFile (single read)

test steps:
- calls `prepare_job`
- creates a dummy `job_context` with one OriginalFile
- calls `salmon._prepare_files` to prepare the OriginalFile before determining the index length
- calls `salmon._determine_index_length`
- asserts that the raw index length is correctly set

> - asserts that the indexl length is correctly set

Reviewer 2: ```suggestion
- asserts that the index length is correctly set
```

### `test_salmon_determine_index_length_double_read`

Tests that the correct index length is calculated when a sample has two OriginalFiles (double read)

test steps:
- calls `prepare_job`
- creates a dummy `job_context` with both OriginalFiles
- calls `salmon._prepare_files` to prepare the OriginalFiles before determining the index length
- calls `salmon._determine_index_length`
- asserts that the raw index length is correctly set

> - asserts that the indexl length is correctly set

Reviewer 2: ```suggestion
- asserts that the index length is correctly set
```

test steps:
- test setup
    - creates test directory
    - creates dummy `job_context`
    - creates dummy Sample
- sets the yml file to one that doesn't exist
- tries to run SalmonTools
- asserts that the job's success is false
- asserts that the job's failure reason is set correctly
- resets the yml file to the correct file

## EarlyTximport Tests

### `test_early_tximport`

Tests that running tximport early works correctly

test steps:
- creates lists of accession codes for incomplete and complete samples
    - incomplete samples has 5 accession codes
    - complete samples has 20 accessioncodes
    - By having 20/25 samples complete, we're just past both the numerical and percent cutoffs for running tximport early
- calls `run_tximport_at_progress_point` to build the `job_context`, create the experiment with samples, and run tximport
- asserts that 'tximported' is in the final `job_context`
- for each sample in the completed samples list:
    - check that the associations were made properly in `create_tximport_job_context`
    - check that the tpm and rds files were created for the sample
    - check that the sample was processed

> - for each sample in the incoplete sample list:

Reviewer 2: ```suggestion
- for each sample in the incomplete sample list:
```
